Add "Find my phone" app.

vchigrin opened the pull request:
This change set adds implementation of client for BLE "Immediate alert service", allowing companion apps get notifications from PineTime watch when requested by user.

Fixes: #343

vchigrin commented:
This pull request is result of joint work of @jmlich and @vchigrin, most of the code is written by Jozef (personal contributions  can be found by commit history).

vchigrin commented:
That is how that app looks like
![photo_2024-04-25_23-57-58](https://github.com/InfiniTimeOrg/InfiniTime/assets/3948754/41c407a7-9db1-465f-8850-b03777c8573f)
After user presses "Mild" or "High" button phone starts ringing and vibrating. User can stop this either by pressing "None" button on the PineTime, or by pressing button in GadgetBridge companion app on the phone.

To make this work some changes in GadgetBridge application were required, I suppose to upstream this after ensuring  that this functionality will be accepted in InfiniTime.

vchigrin commented:
Corresponding PR to GadgetBridge
https://codeberg.org/Freeyourgadget/Gadgetbridge/pulls/3746

NitroNils commented:
And reverse? Find my watch? Make PineTime blink display and vibrate?
Don't know what this would take. Just a logical follow up 😁

jmlich commented:
Find my watch already works. It is Alert Notification Service. In Amazfish -> Settings -> Debug page -> Phone Call.
Also Immediate Alert Service is implemented and shows None/Mild/High levels, but this is not right to find watch.

crazycrystals commented:
Sorry if its offtopic, but is it also possible to have something like a signal strength indicator so you can see how close the device is and triangulate it?
Doing a ring with an app could be broken by do not disturb.

jmlich commented:
The signal strength or more precisely RSSI (Received Signal Strength Indication) is specified/measured in BLE and Nimble can provide the value. I am not sure if it is reasonable reachable in the code.

vchigrin commented on src/FreeRTOSConfig.h:
```diff
@@ -62,7 +62,7 @@
 #define configMINIMAL_STACK_SIZE                (120)
-#define configTOTAL_HEAP_SIZE                   (1024 * 40)
+#define configTOTAL_HEAP_SIZE                   (1024 * 39)
 #define configMAX_TASK_NAME_LEN                 (4)
```
Most probably problem with insufficient memory must be handled in some other way. If this is true, please, tell me how I should do it. With old constant build fails..

mark9064 replied:
Lowering it to 39 is fine, you do not need to worry

vchigrin commented:
Dear maintainers, I kindly ask to look at this PR in the next few months. There is a risk for me to lose access to GitHub in September due to US sanctions package published in June. I am not 100% sure in this scenario - official statements a bit vague and I don't know how GitHub lawyers will interpret them. But, as they say, "hope for the best but prepare for the worst".

devnoname120 commented:
Previous attempt: https://github.com/InfiniTimeOrg/InfiniTime/pull/1193

devnoname120 commented:
FYI here are some notes about the limitations of the `Immediate alert service` standard: https://github.com/InfiniTimeOrg/InfiniTime/pull/1193#issuecomment-1656976536

vchigrin commented:
Well, you have a point. Indeed, using ImmediateAlertService has some drawbacks. Problems you mentioned are not solved here - in this app button returns to original state after timeout in few seconds, there are no any  complex communications. 

So having this PR and your work maintainers have a choice - either use standard BLE service, having some drawbacks, or develop custom solution, which may give better user experience. I'll be happy with any of these choices.

Or may be this functionality is out of "product vision" of the InfiniTime and it will not be merged - I don't know... 

If I can help in any way to bringing this functionality to main branch, let me know, please.

devnoname120 commented:
@vchigrin As far as I see your approach matches better what they were looking for! And no matter what the UI code will be useful. We will have to wait for @JF002 to chime on this 😊

vchigrin commented:
Just noted that corresponding issue at present is the leader by "thumb-up" reactions :)
https://github.com/InfiniTimeOrg/InfiniTime/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

@JF002 @mark9064 I'll be very happy to assist with merging this PR. Is there anything I can do to help with review?

mark9064 commented:
I agree that the BLE spec sucks completely here. Both the IAS and FMP have annoying limitations that makes them kind of useless :(

I think implementing just the IAS is probably our best bet, but we should probably bend the rules a bit and expect the central to keep ringing even on disconnection etc. I honestly have no clue who the IAS is designed for TBH.

As for this PR, I think the UI needs some work. The "None-Mild-High" buttons aren't immediately clear in functionality to the user. Though the specification leaves it vague, I think we could assign more precise meaning to the levels- whether the companion app chooses to do something else doesn't really matter as we don't get to control the meaning of levels anyway. Maybe Stop, Vibrate and Ring? Or maybe we should simplify it to just Stop and Ring (and skip ever sending the mild level)?

(as an aside, I also wanted to remove all client roles from InfiniTime as they're kind of silly but the spec doesn't allow notify either- sending a notification to the central makes sooooo much more sense to me. So that puts a hole in that plan which I'm a little sad about. Oh well)

vchigrin commented:
Updated. I've removed "mild" alert since "Vibrate" label is too long for this button size. If you think 3 buttons will be better, I think we should consider emojis/icons...

At present UI looks like this.
![photo_2024-11-28_01-43-12](https://github.com/user-attachments/assets/aacbffbf-57e6-4b8d-a736-16eeff5e9b62)


PS: There "build-simulator" check fails since simulator repo does not have ImmediateAlertClient class. What is correct way handle situations like this? Should I open PR with necessary  changes to simulator repo beforehand?

mark9064 commented:
I think the two button design looks quite good, what about you?
Not sure if anything should be done with the space above, I'm pretty bad at visual design 😅
Now this PR is feature complete it is a good time to do the InfiniSim PR :+1:

I'll try to review soon but things are looking busy- I'll get to it when I can

vchigrin commented:
Made https://github.com/InfiniTimeOrg/InfiniSim/pull/165

mark9064 commented on src/components/ble/ImmediateAlertClient.cpp:
```diff
@@ -0,0 +1,113 @@
+}
+
+void ImmediateAlertClient::Discover(uint16_t connectionHandle, std::function<void(uint16_t)> onServiceDiscovered) {
+  NRF_LOG_INFO("[IAS] Starting discovery");
```
The log messages would look nicer if they all used `[IAS]` rather than half using `IAS`

vchigrin replied:
Good point, resolved, thanks.

mark9064 reviewed (changes requested):
I think this app could be improved by some UI rework:
- The app should be aware of whether BLE is connected and whether the connected device supports the IAS. If not, it should show an appropriate message and disable the buttons, otherwise they can be enabled. It should also respond to connection changes, so it'll need a `Refresh()` method to check the bluetooth conditions continuously and change the UI as needed
- It should maintain state (only while the app is open) - so it should show a message like "Device alerting" after a successful IAS write, and "Alert stopped" after setting the no alert level. On app load since the IAS state is unknown, it would probably be best to display a simple "Ready"/"Disconnected"/"Alert service unavailable" message and replace this message after sending an alert
- Since signal may be bad when using this service, it should explicitly handle message send failures, and report back a different message to when the service isn't present or BLE is unavailable

Other than those points the rest looks good :+1:

vchigrin commented on src/components/ble/ImmediateAlertClient.h:
```diff
@@ -0,0 +1,61 @@
+      bool OnDiscoveryEvent(uint16_t connectionHandle, const ble_gatt_error* error, const ble_gatt_svc* service);
+      int OnCharacteristicDiscoveryEvent(uint16_t conn_handle, const ble_gatt_error* error, const ble_gatt_chr* characteristic);
+
+      bool sendImmediateAlert(Levels level);
```
Done.

vchigrin commented on src/displayapp/screens/FindMyPhone.cpp:
```diff
@@ -0,0 +1,118 @@
+}
+
+FindMyPhone::FindMyPhone(Pinetime::Controllers::ImmediateAlertClient& immediateAlertClient) : immediateAlertClient {immediateAlertClient} {
+  last_level = Pinetime::Controllers::ImmediateAlertClient::Levels::NoAlert;
```
Done, thanks.

vchigrin commented on src/components/ble/ImmediateAlertClient.cpp:
```diff
@@ -0,0 +1,113 @@
+       .uuid = &immediateAlertClientUuid.u,
+       .characteristics = characteristicDefinition},
+      {0},
+    } {
```
Done, thanks

vchigrin commented on .gitignore:
```diff
@@ -7,6 +7,7 @@ cmake-build-*
 cmake-*/
 CMakeFiles
 **/CMakeCache.txt
+CMakeLists.txt.user*
```
Seems this is an accident - removed, thanks.

vchigrin commented on src/components/ble/ImmediateAlertClient.cpp:
```diff
@@ -0,0 +1,113 @@
+                                                         const ble_gatt_chr* characteristic) {
+
+  if (error->status != 0 && error->status != BLE_HS_EDONE) {
+    NRF_LOG_INFO("IAS Characteristic discovery ERROR");
```
One is when bluetooth lib returns error to us and other when no error occured, just characteristic not found. Re-worked messages to make it more clear.

vchigrin commented on src/components/ble/ImmediateAlertClient.cpp:
```diff
@@ -0,0 +1,113 @@
+    return false;
+  }
+
+  ble_gattc_write_no_rsp(connectionHandle, alertLevelHandle, om);
```
Thank you so much for spotting this. Changed a bit this class, used std::optional to making using non-initialized values a bit harder (programmer will notice that checking for non-null optional should be performed).

vchigrin commented on src/displayapp/screens/FindMyPhone.cpp:
```diff
@@ -0,0 +1,118 @@
+  lv_obj_align(bt_none, nullptr, LV_ALIGN_IN_BOTTOM_LEFT, 0, 0);
+  label_none = lv_label_create(bt_none, nullptr);
+  lv_label_set_text_static(label_none, noneLabelText);
+  lv_obj_set_style_local_bg_color(bt_none, LV_LABEL_PART_MAIN, LV_STATE_DEFAULT, Colors::lightGray);
```
Done. Also renamed buttons to `btnStop` and `btnRing` to make it more consistent with UI labels.

vchigrin commented on src/displayapp/screens/FindMyPhone.cpp:
```diff
@@ -0,0 +1,118 @@
+  label_title = lv_label_create(lv_scr_act(), nullptr);
+
+  lv_label_set_text_static(label_title, defaultLabelText);
+  lv_obj_set_style_local_text_color(label_title, LV_LABEL_PART_MAIN, LV_STATE_DEFAULT, Colors::lightGray);
```
Done.

vchigrin commented on src/displayapp/screens/FindMyPhone.cpp:
```diff
@@ -0,0 +1,118 @@
+  static constexpr char noConnectionLabelText[] = "No connection";
+  static constexpr char noneLabelText[] = "Stop";
+  static constexpr char highLabelText[] = "Ring";
+  static constexpr auto restoreLabelTimeoutTicks = pdMS_TO_TICKS(2 * 1000);
```
Removed this constant as UI logic changed a lot. Following you suggestion UI now tracks BLE disconnect and send failures.

vchigrin commented on src/displayapp/screens/FindMyPhone.cpp:
```diff
@@ -0,0 +1,118 @@
+      break;
+  }
+  if (immediateAlertClient.sendImmediateAlert(last_level)) {
+    lv_label_set_text_static(label_title, alertSentLabelText);
```
UI was changed a lot. Now there is separate constants describing label state, in that case see `alertingLabelState` and `stoppedLabelState` constants. Thank you for the suggestion.

mark9064 commented:
Thank you for all the changes, I will try to review them properly soon! Apologies in advance if it takes me a while to get round to it

joergi commented:
Looking forward for this feature ♥️
